File: frontend/shared/protocol/message/drequest.ts
// Copyright 2023 The Z2K Plus+ Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {MetadataRecord, SearchOrigin, ZgramCore, ZgramId} from "../zephyrgram";
import {
  assertAndDestructure1,
  assertAndDestructure2,
  assertAndDestructure3,
  assertAndDestructure4,
  assertArray,
  assertBoolean,
  assertEnum,
  assertNumber,
  assertString, optionalToJson, optionalTryParseJson
}
  from "../../json_util";
import {intercalate, makeCommaSeparatedList, Optional, Pair, staticFail} from "../../utility";
import {Filter} from "../misc";

export namespace drequests {
  export class CheckSyntax {
    constructor(readonly query: string) {
    }

    toJson() {
      return [this.query];
    }

    static tryParseJson(item: any) {
      const [query] = assertAndDestructure1(item, assertString);
      return new CheckSyntax(query);
    }

    toString() {
      return `CheckSyntax(${this.query})`;
    }
  }

  export class Subscribe {
    constructor(readonly query: string, readonly searchOrigin: SearchOrigin, readonly pageSize: number,
        readonly queryMargin: number) {
    }

    toJson() {
      return [this.query, this.searchOrigin.toJson(), this.pageSize, this.queryMargin];
    }

    static tryParseJson(item: any) {
      const [q, so, ps, qm] = assertAndDestructure4(item, assertString, SearchOrigin.tryParseJson,
          assertNumber, assertNumber);
      return new Subscribe(q, so, ps, qm);
    }

    toString() {
      return `Subscribe(${this.query}, ${this.searchOrigin}, ${this.pageSize}, ${this.queryMargin}`;
    }
  }

  export class GetMoreZgrams {
    constructor(public readonly forBackSide: boolean, public readonly count: number) {
    }

    toJson() {
      return [this.forBackSide, this.count];
    }

    static tryParseJson(item: any) {
      const [bs, c] = assertAndDestructure2(item, assertBoolean, assertNumber);
      return new GetMoreZgrams(bs, c);
    }

    toString() {
      return `GetMoreZgrams(${this.forBackSide}, ${this.count})`;
    }
  }

  export class PostZgrams {
    constructor(readonly zgramsAndRefersTo: Pair<ZgramCore, Optional<ZgramId>>[]) {
    }

    toJson() {
      const zgJson = this.zgramsAndRefersTo.map(zgPair => zgPair.toJson());
      return [zgJson];
    }

    static tryParseJson(item: any) {
      const array = assertAndDestructure1(item, assertArray);
      const pairs = array.map(
          pair => Pair.tryParseJson(
              pair,
              first => ZgramCore.tryParseJson(first),
              second => Optional.tryParseJson(second, ZgramId.tryParseJson))
      );
      return new PostZgrams(pairs);
    }

    toString() {
      return `Post(${makeCommaSeparatedList(this.zgramsAndRefersTo)})`;
    }
  }

  export class GetSpecificZgrams {
    constructor(readonly zgramIds: ZgramId[]) {
    }

    toJson() {
      const zgJson = this.zgramIds.map(zgId => zgId.toJson());
      return [zgJson];
    }

    static tryParseJson(item: any) {
      const array = assertAndDestructure1(item, assertArray);
      const zgIds = array.map(ZgramId.tryParseJson);
      return new GetSpecificZgrams(zgIds);
    }

    toString() {
      return `Post(${makeCommaSeparatedList(this.zgramIds)})`;
    }
  }

  export class PostMetadata {
    constructor(readonly metadata: MetadataRecord[]) {
    }

    toJson() {
      const mdJson = this.metadata.map(o => o.toJson());
      return [mdJson];
    }

    static tryParseJson(item: any) {
      const [mdArray] = assertAndDestructure1(item, assertArray);
      const mds = mdArray.map(MetadataRecord.tryParseJson);
      return new PostMetadata(mds);
    }

    toString() {
      return `Post(${makeCommaSeparatedList(this.metadata)})`;
    }
  }

  export class ProposeFilters {
    constructor(public readonly basedOnVersion: number, public readonly theseFiltersAreNew: boolean,
        public readonly filters: Filter[]) {
    }

    toJson() {
      return [
        this.basedOnVersion,
        this.theseFiltersAreNew,
        this.filters.map(f => f.toJson())
      ];
    }

    static tryParseJson(item: any) {
      const [bov, fan, fs] = assertAndDestructure3(item,
          assertNumber, assertBoolean, assertArray);
      const filters = fs.map(Filter.tryParseJson);
      return new ProposeFilters(bov, fan, filters);
    }

    toString() {
      var filtersList = intercalate(", ", f => f.toString(), this.filters);
      return `Filter(${this.basedOnVersion}, ${this.theseFiltersAreNew}, ${filtersList})`;
    }
  }


export class Ping {
    constructor(readonly cookie: number) {
    }

    toJson() {
      return [this.cookie];
    }

    static tryParseJson(item: any) {
      const [cookie] = assertAndDestructure1(item, assertNumber);
      return new Ping(cookie);
    }

    toString() {
      return `Ping(${this.cookie})`;
    }
  }
}

namespace drequestInfo {
  export const enum Tag {
    // Check the syntax of a query
    CheckSyntax = "CheckSyntax",
    // Subscribe to or unsubscribe from a stream of Zgrams
    Subscribe = "Subscribe",
    // Request (up to) N more zgrams from the front or back side.
    GetMoreZgrams = "GetMoreZgrams",
    // Post new zgrams
    PostZgrams = "PostZgrams",
    // Post new metadata
    PostMetadata = "PostMetadata",
    // Get specific zgrams that you happen to care about (e.g. those that appear in a refers-to)
    GetSpecificZgrams = "GetSpecificZgrams",
    // Propose new filters
    ProposeFilters = "ProposeFilters",
    // For debugging. A round-trip to the server. When you get the Ack back, you know that the
    // server has processed all your requests up through the Ack.
    Ping = "Ping",
  }

  export type payload_t = drequests.CheckSyntax | drequests.Subscribe |
      drequests.GetMoreZgrams | drequests.PostZgrams | drequests.PostMetadata |
      drequests.GetSpecificZgrams | drequests.ProposeFilters | drequests.Ping;

  export interface ICheckSyntax {
    readonly tag: Tag.CheckSyntax;
    readonly payload: drequests.CheckSyntax;
  }

  export interface ISubscribe {
    readonly tag: Tag.Subscribe;
    readonly payload: drequests.Subscribe;
  }

  export interface IGetMoreZgrams {
    readonly tag: Tag.GetMoreZgrams;
    readonly payload: drequests.GetMoreZgrams;
  }

  export interface IPostZgrams {
    readonly tag: Tag.PostZgrams;
    readonly payload: drequests.PostZgrams;
  }

  export interface IPostMetadata {
    readonly tag: Tag.PostMetadata;
    readonly payload: drequests.PostMetadata;
  }

  export interface IGetSpecificZgrams {
    readonly tag: Tag.GetSpecificZgrams;
    readonly payload: drequests.GetSpecificZgrams;
  }

  export interface IProposeFilters {
    readonly tag: Tag.ProposeFilters;
    readonly payload: drequests.ProposeFilters;
  }

  export interface IPing {
    readonly tag: Tag.Ping;
    readonly payload: drequests.Ping;
  }

  export interface IVisitor<TResult> {
    visitCheckSyntax(o: drequests.CheckSyntax): TResult;

    visitSubscribe(o: drequests.Subscribe): TResult;

    visitGetMoreZgrams(o: drequests.GetMoreZgrams): TResult;

    visitPostZgrams(o: drequests.PostZgrams): TResult;

    visitGetSpecificZgrams(o: drequests.GetSpecificZgrams): TResult;

    visitPostMetadata(o: drequests.PostMetadata): TResult;

    visitProposeFilters(o: drequests.ProposeFilters): TResult;

    visitPing(o: drequests.Ping): TResult;
  }
}  // namespace drequestInfo

export type IDRequest = drequestInfo.ICheckSyntax | drequestInfo.ISubscribe |
    drequestInfo.IGetMoreZgrams | drequestInfo.IPostZgrams | drequestInfo.IGetSpecificZgrams |
    drequestInfo.IPostMetadata | drequestInfo.IProposeFilters | drequestInfo.IPing;

export class DRequest {
  static createCheckSyntax(query: string) {
    const payload = new drequests.CheckSyntax(query);
    return new DRequest(drequestInfo.Tag.CheckSyntax, payload);
  }
  static createSubscribe(query: string, searchOrigin: SearchOrigin, pageSize: number, queryMargin: number) {
    const payload = new drequests.Subscribe(query, searchOrigin, pageSize, queryMargin);
    return new DRequest(drequestInfo.Tag.Subscribe, payload);
  }
  static createGetMoreZgrams(forBackSide: boolean, count: number) {
    return new DRequest(drequestInfo.Tag.GetMoreZgrams, new drequests.GetMoreZgrams(forBackSide, count));
  }
  static createPostZgrams(zgPairs: Pair<ZgramCore, Optional<ZgramId>>[]) {
    return new DRequest(drequestInfo.Tag.PostZgrams, new drequests.PostZgrams(zgPairs));
  }
  static createGetSpecificZgrams(zgramIds: ZgramId[]) {
    return new DRequest(drequestInfo.Tag.GetSpecificZgrams, new drequests.GetSpecificZgrams(zgramIds));
  }
  static createPostMetadata(metadata: MetadataRecord[]) {
    return new DRequest(drequestInfo.Tag.PostMetadata, new drequests.PostMetadata(metadata));
  }
  static createProposeFilters(basedOnVersion: number, theseFiltersAreNew: boolean, filters: Filter[]) {
    return new DRequest(drequestInfo.Tag.ProposeFilters, new drequests.ProposeFilters(basedOnVersion,
        theseFiltersAreNew, filters));
  }
  static createPing(cookie: number) {
    return new DRequest(drequestInfo.Tag.Ping, new drequests.Ping(cookie));
  }

  constructor(readonly tag : drequestInfo.Tag, readonly payload: drequestInfo.payload_t) {
  }
  
  acceptVisitor<TResult>(visitor: drequestInfo.IVisitor<TResult>) : TResult {
    const idreq = this as IDRequest;
    switch (idreq.tag) {
      case drequestInfo.Tag.CheckSyntax: return visitor.visitCheckSyntax(idreq.payload);
      case drequestInfo.Tag.Subscribe: return visitor.visitSubscribe(idreq.payload);
      case drequestInfo.Tag.GetMoreZgrams: return visitor.visitGetMoreZgrams(idreq.payload);
      case drequestInfo.Tag.PostZgrams: return visitor.visitPostZgrams(idreq.payload);
      case drequestInfo.Tag.PostMetadata: return visitor.visitPostMetadata(idreq.payload);
      case drequestInfo.Tag.GetSpecificZgrams: return visitor.visitGetSpecificZgrams(idreq.payload);
      case drequestInfo.Tag.ProposeFilters: return visitor.visitProposeFilters(idreq.payload);
      case drequestInfo.Tag.Ping: return visitor.visitPing(idreq.payload);
      default: staticFail(idreq);
    }
    throw new Error("notreached");
  }

  toJson() {
    const variant = [this.tag, this.payload.toJson()];
    return [variant];
  }

  toString() {
    return `Request(${this.tag}, ${this.payload})`;
  }

  static tryParseJson(item: any) {
    const [variant] = assertAndDestructure1(item, assertArray);
    const [tag, item1] = assertAndDestructure2(variant, i => assertEnum<drequestInfo.Tag>(i), i => i);
    let payload: drequestInfo.payload_t;
    switch (tag) {
      case drequestInfo.Tag.CheckSyntax: payload = drequests.CheckSyntax.tryParseJson(item1); break;
      case drequestInfo.Tag.Subscribe: payload = drequests.Subscribe.tryParseJson(item1); break;
      case drequestInfo.Tag.GetMoreZgrams: payload = drequests.GetMoreZgrams.tryParseJson(item1); break;
      case drequestInfo.Tag.PostZgrams: payload = drequests.PostZgrams.tryParseJson(item1); break;
      case drequestInfo.Tag.PostMetadata: payload = drequests.PostMetadata.tryParseJson(item1); break;
      case drequestInfo.Tag.GetSpecificZgrams: payload = drequests.GetSpecificZgrams.tryParseJson(item1); break;
      case drequestInfo.Tag.ProposeFilters: payload = drequests.ProposeFilters.tryParseJson(item1); break;
      case drequestInfo.Tag.Ping: payload = drequests.Ping.tryParseJson(item1); break;
      default: throw staticFail(tag);
    }
    return new DRequest(tag, payload);
  }
}

File: backend/src/shared/protocol/message/drequest.cc
// Copyright 2023 The Z2K Plus+ Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include "z2kplus/backend/shared/protocol/message/drequest.h"
#include "kosak/coding/coding.h"
#include "kosak/coding/failures.h"
#include "kosak/coding/myjson.h"

using kosak::coding::ParseContext;
using kosak::coding::streamf;

namespace z2kplus::backend::shared::protocol::message {

namespace drequests {

CheckSyntax::CheckSyntax() = default;
CheckSyntax::CheckSyntax(std::string query) : query_(std::move(query)) {}
CheckSyntax::CheckSyntax(CheckSyntax &&) noexcept = default;
CheckSyntax &CheckSyntax::operator=(CheckSyntax &&) noexcept = default;
CheckSyntax::~CheckSyntax() = default;

std::ostream &operator<<(std::ostream &s, const CheckSyntax &o) {
  return streamf(s, "CheckSyntax(%o)", o.query_);
}

DEFINE_TYPICAL_JSON(CheckSyntax, query_);

Subscribe::Subscribe() = default;
Subscribe::Subscribe(std::string &&query, SearchOrigin start, size_t pageSize, size_t queryMargin) :
    query_(std::move(query)), start_(start), pageSize_(pageSize), queryMargin_(queryMargin) {}
Subscribe::Subscribe(Subscribe &&) noexcept = default;
Subscribe &Subscribe::operator=(Subscribe &&) noexcept = default;
Subscribe::~Subscribe() = default;

std::ostream &operator<<(std::ostream &s, const Subscribe &o) {
  return streamf(s, "Subscribe(%o,%o,%o,%o)", o.query_, o.start_, o.start_, o.pageSize_,
      o.queryMargin_);
}

DEFINE_TYPICAL_JSON(Subscribe, query_, start_, pageSize_, queryMargin_);

GetMoreZgrams::GetMoreZgrams() = default;
GetMoreZgrams::GetMoreZgrams(bool forBackSide, uint64_t count) : forBackSide_(forBackSide),
    count_(count) {}
GetMoreZgrams::GetMoreZgrams(GetMoreZgrams &&) noexcept = default;
GetMoreZgrams &GetMoreZgrams::operator=(GetMoreZgrams &&) noexcept = default;
GetMoreZgrams::~GetMoreZgrams() = default;

std::ostream &operator<<(std::ostream &s, const GetMoreZgrams &o) {
  return streamf(s, "GetMore(%o,%o)", o.forBackSide_, o.count_);
}

DEFINE_TYPICAL_JSON(GetMoreZgrams, forBackSide_, count_);

PostZgrams::PostZgrams() = default;
PostZgrams::PostZgrams(std::vector<entry_t> entries) : entries_(std::move(entries)) {}
PostZgrams::PostZgrams(PostZgrams &&) noexcept = default;
PostZgrams &PostZgrams::operator=(PostZgrams &&) noexcept = default;
PostZgrams::~PostZgrams() = default;

std::ostream &operator<<(std::ostream &s, const PostZgrams &o) {
  return streamf(s, "PostZgrams(entries=%o)", o.entries_);
}
DEFINE_TYPICAL_JSON(PostZgrams, entries_);

PostMetadata::PostMetadata() = default;
PostMetadata::PostMetadata(std::vector<MetadataRecord> metadata) : metadata_(std::move(metadata)) {}
PostMetadata::PostMetadata(PostMetadata &&) noexcept = default;
PostMetadata &PostMetadata::operator=(PostMetadata &&) noexcept = default;
PostMetadata::~PostMetadata() = default;

std::ostream &operator<<(std::ostream &s, const PostMetadata &o) {
  return streamf(s, "PostMetadata(md=%o)", o.metadata_);
}
DEFINE_TYPICAL_JSON(PostMetadata, metadata_);

GetSpecificZgrams::GetSpecificZgrams() = default;
GetSpecificZgrams::GetSpecificZgrams(std::vector<ZgramId> zgramIds) : zgramIds_(std::move(zgramIds)) {}
GetSpecificZgrams::GetSpecificZgrams(GetSpecificZgrams &&) noexcept = default;
GetSpecificZgrams &GetSpecificZgrams::operator=(GetSpecificZgrams &&) noexcept = default;
GetSpecificZgrams::~GetSpecificZgrams() = default;

std::ostream &operator<<(std::ostream &s, const GetSpecificZgrams &o) {
  return streamf(s, "GetSpecificZgrams(zgramIds=%o)", o.zgramIds_);
}
DEFINE_TYPICAL_JSON(GetSpecificZgrams, zgramIds_);

ProposeFilters::ProposeFilters() = default;
ProposeFilters::ProposeFilters(uint64_t basedOnVersion, bool theseFiltersAreNew, std::vector<Filter> filters) :
  basedOnVersion_(basedOnVersion), theseFiltersAreNew_(theseFiltersAreNew), filters_(std::move(filters)) {}
ProposeFilters::ProposeFilters(ProposeFilters &&) noexcept = default;
ProposeFilters &ProposeFilters::operator=(ProposeFilters &&) noexcept = default;
ProposeFilters::~ProposeFilters() = default;

std::ostream &operator<<(std::ostream &s, const ProposeFilters &o) {
  return streamf(s, "ProposeFilters(%o, %o, %o)", o.basedOnVersion_, o.theseFiltersAreNew_, o.filters_);
}
DEFINE_TYPICAL_JSON(ProposeFilters, basedOnVersion_, theseFiltersAreNew_, filters_);

std::ostream &operator<<(std::ostream &s, const Ping &o) {
  return streamf(s, "Ping(%o)", o.cookie_);
}
DEFINE_TYPICAL_JSON(Ping, cookie_);

DEFINE_VARIANT_JSON(DRequestPayloadHolder, drequestPayload_t);
}  // namespace drequests

DRequest::DRequest() = default;
DRequest::DRequest(drequests::CheckSyntax o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::Subscribe o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::GetMoreZgrams o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::PostZgrams o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::PostMetadata o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::GetSpecificZgrams o) : payload_(std::move(o)) {}
DRequest::DRequest(drequests::Ping o) : payload_(std::move(o)) {}
DRequest::DRequest(DRequest &&) noexcept = default;
DRequest &DRequest::operator=(DRequest &&) noexcept = default;
DRequest::~DRequest() = default;

DEFINE_TYPICAL_JSON(DRequest, payload_);
}  // namespace z2kplus::backend::shared::protocol
